cmd/sharded-test-server/virtual.go

```diff
@@ -19,70 +19,239 @@ package main
+		args:        args,
+	}, nil
+}
+
+func (v *VirtualWorkspace) start(ctx context.Context) error {
+	prefix := fmt.Sprintf("VW-%d", v.index)
 	yellow := color.New(color.BgYellow, color.FgHiWhite).SprintFunc()
 	out := lineprefix.New(
 		lineprefix.Prefix(yellow(prefix)),
 		lineprefix.Color(color.New(color.FgHiYellow)),
 	)
 
-	authenticationKubeconfigPath := filepath.Join(workDirPath, fmt.Sprintf(".kcp-%d", index), "admin.kubeconfig")
-	clientCAFilePath := filepath.Join(workDirPath, ".kcp", "client-ca.crt")
-
 	commandLine := framework.DirectOrGoRunCommand("virtual-workspaces")
+	commandLine = append(commandLine, v.args...)
 	commandLine = append(
 		commandLine,
-		fmt.Sprintf("--kubeconfig=%s", authenticationKubeconfigPath),
-		fmt.Sprintf("--authentication-kubeconfig=%s", authenticationKubeconfigPath),
 		"--authentication-skip-lookup",
-		fmt.Sprintf("--client-ca-file=%s", clientCAFilePath),
-		fmt.Sprintf("--tls-private-key-file=%s", filepath.Join(workDirPath, ".kcp/serving-ca.key")),
-		fmt.Sprintf("--tls-cert-file=%s", filepath.Join(workDirPath, ".kcp/serving-ca.crt")),
-		fmt.Sprintf("--secure-port=%d", 7444+index),
 		"--requestheader-username-headers=X-Remote-User",
 		"--requestheader-group-headers=X-Remote-Group",
-		fmt.Sprintf("--requestheader-client-ca-file=%s", filepath.Join(workDirPath, ".kcp/requestheader-ca.crt")),
+		fmt.Sprintf("--requestheader-client-ca-file=%s", filepath.Join(v.workDirPath, ".kcp/requestheader-ca.crt")),
 		"--v=4",
 	)
 	fmt.Fprintf(out, "running: %v\n", strings.Join(commandLine, " "))
 
 	cmd := exec.CommandContext(ctx, commandLine[0], commandLine[1:]...) //nolint:gosec
 
-	logFilePath := filepath.Join(logDirPath, fmt.Sprintf("kcp-virtual-workspaces-%d.log", index))
+	logFilePath := filepath.Join(v.workDirPath, fmt.Sprintf(".kcp-virtual-workspaces-%d/virtualworkspace.log", v.index))
+	if v.logDirPath != "" {
+		logFilePath = filepath.Join(v.logDirPath, fmt.Sprintf("kcp-virtual-workspaces-%d.log", v.index))
+	}
+
 	if err := os.MkdirAll(filepath.Dir(logFilePath), 0755); err != nil {
-		return nil, err
+		return err
 	}
 	logFile, err := os.OpenFile(logFilePath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	writer := helpers.NewHeadWriter(logFile, out)
-	cmd.Stdout = writer
+	v.writer = helpers.NewHeadWriter(logFile, out)
+	cmd.Stdout = v.writer
 	cmd.Stdin = os.Stdin
-	cmd.Stderr = writer
+	cmd.Stderr = v.writer
 
 	if err := cmd.Start(); err != nil {
-		return nil, err
+		return err
 	}
 
 	terminatedCh := make(chan error, 1)
+	v.terminatedCh = terminatedCh
 	go func() {
 		terminatedCh <- cmd.Wait()
 	}()
 
-	return terminatedCh, nil
+	return nil
+}
+
+func (v *VirtualWorkspace) waitForReady(ctx context.Context) (<-chan error, error) {
+	// wait for readiness
+	logger := klog.FromContext(ctx)
+	logger.WithValues("virtual-workspaces", v.index).Info("Waiting for virtual-workspaces /readyz to succeed")
+	for {
+		time.Sleep(100 * time.Millisecond)
```

Comment:
nit: prefer a `time.Ticker()` or any of the `wait.*` library functions for convenience.

Reply:
They do things like handle cancellation contexts for you etc

Reply:
Done